File: Java/src/test/java/de/steffen_liersch/calculator/CalculatorTests.java
/*--------------------------------------------------------------------------*\
::
::  Copyright © 2023 Steffen Liersch
::  https://www.steffen-liersch.de/
::
\*--------------------------------------------------------------------------*/

package de.steffen_liersch.calculator;

import com.fasterxml.jackson.databind.JsonNode;
import org.junit.Assert;
import org.junit.Test;

import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;

public final class CalculatorTests
{
  @Test
  public void performTests()
  {
    int testCount = 0;
    int errorCount = 0;
    var calculator = new Calculator(new FloatFormatter(16));
    String json = Helpers.loadTests();
    JsonNode tests = JSON.parse(json);
    for(JsonNode test : tests)
    {
      System.out.println("Test: " + test.get("name").asText());
      List<String> output = toStringArray(test.get("output"));
      List<String> array = toStringArray(test.get("expression"));
      for(String expr : array)
      {
        List<String> lines = calculator.calculateAndFormat(expr, "");
        if(!assertEqual(output, lines))
          errorCount++;
        testCount++;
      }
    }
    System.out.println(
        MessageFormat.format("=> {0} tests completed with {1} errors", testCount, errorCount));

    if(errorCount > 0)
      Assert.fail();
  }

  private static boolean assertEqual(List<String> expected, List<String> actual)
  {
    int c1 = expected.size();
    int c2 = actual.size();
    if(c1 != c2)
    {
      System.out.println("Assertion failed: Arrays with different lengths");
      return false;
    }

    for(int i = 0; i < c1; i++)
    {
      if(!expected.get(i).equals(actual.get(i)))
      {
        System.out.println(
            MessageFormat.format("Assertion failed ({0}, {1}", expected.get(i), actual.get(i)));
        return false;
      }
    }

    return true;
  }

  private static List<String> toStringArray(JsonNode node)
  {
    var res = new ArrayList<String>();
    if(!node.isArray())
      res.add(node.asText());
    else
    {
      for(var x : node)
        res.add(x.asText());
    }

    return res;
  }
}

File: ObjectPascal/CalculatorTests.pas
(*--------------------------------------------------------------------------*]
::
::  Copyright © 2023 Steffen Liersch
::  https://www.steffen-liersch.de/
::
[*--------------------------------------------------------------------------*)

{$I Options.inc}

unit CalculatorTests;

interface

uses
  fpjson,
  SysUtils,
  Calculator, FloatFormatter;

function PerformTests(Tests: TJSONArray): boolean;

implementation

function AssertEqual(Expected: TJSONArray; const Actual: TCalculatorStrings): boolean;
var
  C1, C2, I: integer;
begin
  C1 := Expected.Count;
  C2 := Length(Actual);
  if C1 <> C2 then
  begin
    WriteLn('Assertion failed: Arrays with different lengths');
    Exit(false);
  end;

  for I := 0 to C1 - 1 do
  begin
    if Expected.Strings[I] <> Actual[I] then
    begin
      WriteLn('Assertion failed', Expected.Strings[I], Actual[I]);
      Exit(false);
    end;
  end;

  Exit(true);
end;

function PerformTests(Tests: TJSONArray): boolean;
var
  Calculator: TCalculator;
  Formatter: TFloatFormatter;
  TestCount, ErrorCount, I, J: integer;
  Test: TJSONObject;
  Output, Expr: TJSONArray;
  Lines: TCalculatorStrings;
begin
  TestCount := 0;
  ErrorCount := 0;
  Formatter.Create(16);
  Calculator := TCalculator.Create(Formatter);
  try
    for I := 0 to Tests.Count - 1 do
    begin
      Test := Tests[I] as TJSONObject;
      WriteLn('Test: ' + Test.Strings['name']);
      Output := Test.Arrays['output'];
      Expr := Test.Arrays['expression'];
      for J := 0 to Expr.Count - 1 do
      begin
        Lines := calculator.CalculateAndFormat(Expr.Strings[J], '');
        if not AssertEqual(Output, Lines) then
          Inc(ErrorCount);
        Inc(TestCount);
      end;
    end;
  finally
    Calculator.Free;
  end;
  WriteLn(Format('=> %d tests completed with %d errors', [TestCount, ErrorCount]));
  Result := ErrorCount <= 0;
end;

end.
